src/aspire/source/coordinates.py

```diff
@@ -0,0 +1,485 @@
+        first_mrc = self.mrc_paths[first_mrc_index]
+        with mrcfile.open(first_mrc) as mrc_file:
+            mrc_dtype = np.dtype(mrc_file.data.dtype)
+            shape = mrc_file.data.shape
+        if len(shape) != 2:
+            raise ValueError(
+                "Shape of mrc file is {shape} but expected shape of size 2. Are these unaligned micrographs?"
+            )
+        if self.dtype != mrc_dtype:
+            logger.warning(
+                f"dtype of micrograph is {mrc_dtype}. Will attempt to cast to {self.dtype}"
+            )
+
+        # save the shape to compare the rest of the mrcs against
+        self.mrc_shape = shape
+
+        # look at first coord to get particle size
+        # here we're checking the final coordinate of the first particle
+        # which is the Y-size of the box (the same as the X-size)
+        L = first_coord[3]
+
+        # total number of particles given in coord files
+        # before removing boundary particles and satisfying max_rows
+        original_n = len(self.particles)
+
+        logger.info(f"Micrograph size = {self.mrc_shape[1]}x{self.mrc_shape[0]}")
+        logger.info(f"Particle size = {L}x{L}")
+        self._original_resolution = L
+
+        # remove particles whose boxes do not fit at given particle_size
+        # and get number removed
+        boundary_removed = self.exclude_boundary_particles()
+        # if max_rows is specified, only return up to max_rows many
+        # (after excluding boundary particles)
+        if max_rows:
+            max_rows = min(max_rows, original_n - boundary_removed)
+            self.particles = self.particles[:max_rows]
+
+        # final number of particles in *this* source
+        n = len(self.particles)
+        # total micrographs and particles represented by source (info)
+        logger.info(
+            f"CoordinateSource from {os.path.dirname(first_mrc)} contains {len(mrc_paths)} micrographs, {original_n} picked particles."
+        )
+        # total particles we can load given particle_size (info)
+        if boundary_removed > 0:
+            logger.info(
+                f"{boundary_removed} particles did not fit into micrograph dimensions at particle size {L}, so were excluded"
+            )
+            logger.info(
+                f"Maximum number of particles at this particle size is {original_n - boundary_removed}."
+            )
+        # total particles loaded (specific to this instance)
+        logger.info(f"CoordinateSource object contains {n} particles.")
+
+        ImageSource.__init__(self, L=L, n=n, dtype=dtype)
+
+        # Create filter indices for the source. These are required in order to
+        # pass through the filter eval code.
+        # Bypassing the filter_indices setter in ImageSource allows us
+        # create this source with absolutely *no* metadata.
+        # Otherwise, six default Relion columns are created w/defualt values
```

Comment:
Yes, I believe this can change once the other PR is in, but it also won't affect anything if it stays for now. Could try removing it once both PRs are there?

Reply:
Of course, np.